### PR TITLE
Setup API Data Fetching Architecture

### PR DESCRIPTION
**Pull request by nickjingyuli**
# Description & Technical Solution

Setup API Data Fetching Architecture, including:
* Created `PdfPropData` class, responsible for holding all page data needed for the PDFs and Propositions.
* Setup `reducer` and `context` needed for components state data(`PdfPropData`).
* Setup route handlers, which used by client components to fetch data. See usage example in `library -> propositions -> page.tsx`
* Currently not connected to BE API, returning stub data at the moment.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] Already rebased against main branch.

# Screenshots
Screenshot 2024-02-01 at 4 00 06 PM

Provide screenshots or videos of the changes made if any.

### CODE REVIEW

File: src/lib/interfaces.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      res.push({
+        title: val.proposition_type,
+        mode: mode,
+        isSaved: val.is_saved,
+        content: val.description,
+      });
+    });
+    return res;
+  }
+}
```
Comment (nickjingyuli) [inline src/lib/interfaces.ts:62]: Add methods to get the data needed for your page!

File: src/lib/context/data.context.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      <DataDispatchContext.Provider value={dispatch}>
+        {children}
+      </DataDispatchContext.Provider>
+    </DataContext.Provider>
+  );
+}
+
+// Reducer
+function dataReducer(data: PdfPropData, action: PdfPropDataAction) {
+  switch (action.type) {
```
Comment (nickjingyuli) [inline src/lib/context/data.context.tsx:41]: Only 3 actions added. Need to add the rest.

File: src/app/api/proposition/endpoints.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  });
+}
+
+export async function getLibraryPropositions(
+  uid: string
+): Promise<{ propositions: LibraryPropositionResponse[] }> {
+  return await new Promise((resolve, reject) => {
+    setTimeout(() => {
+      resolve({
+        propositions: [
```
Comment (nickjingyuli) [inline src/app/api/proposition/endpoints.ts:35]: Returning stub data at the moment

**Review by ShuangDa1018 — APPROVED** [review]
ok

**Pull request merged**